plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php

```diff
@@ -451,9 +604,46 @@ public function get_description(): string {
 	 * @param WP_Post $post The deleted post.
 	 */
 	private function handle_deleted_post( $postid, $post ): void {
-		if ( 'shop_order' === $post->post_type && $this->data_sync_is_enabled() ) {
+		global $wpdb;
+
+		if ( 'shop_order' !== $post->post_type && 'shop_order_refund' !== $post->post_type ) {
+			return;
+		}
+
+		$features_controller = wc_get_container()->get( FeaturesController::class );
+		$feature_is_enabled  = $features_controller->feature_is_enabled( 'custom_order_tables' );
+		if ( ! $feature_is_enabled ) {
+			return;
+		}
+
+		if ( $this->data_sync_is_enabled() ) {
 			$this->data_store->delete_order_data_from_custom_order_tables( $postid );
+		} elseif ( $this->custom_orders_table_is_authoritative() ) {
+			return;
+		}
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.SlowDBQuery
+		if ( $wpdb->get_var(
```

Comment:
Seems like the first check will always be false when `$this->data_sync_is_enabled()` is true, and if it is false and posts are authoritative, we only need the first check and then insert directly, wdyt?

Reply:
By "the checks" you mean the `SELECT EXISTS` and the `AND NOT EXSITS`? The second one is there to prevent accidental creation of duplicat records.

Reply:
Hmm, I was thinking along the lines of if we duplicate deleted_from records, would it really be a problem, but this check seems pretty lightweight, so it should be fine.